core/ipc/twin/common/src/main/java/org/opennms/core/ipc/twin/common/AbstractTwinSubscriber.java

```diff
@@ -30,177 +30,83 @@
-            // We can't apply patch when the version jumps more than one version.
-            if (twinUpdate.getVersion() != twinTracker.getVersion() + 1) {
-                sendRpcRequest(new TwinRequest(twinUpdate.getKey(), twinUpdate.getLocation()));
-                return;
-            }
-            byte[] patchedBytes = applyPatch(twinTracker, twinUpdate);
-            // Invoke RPC when we can't apply patch properly.
-            if (patchedBytes != null) {
-                // Update twin object for sessions.
-                updateSessions(twinUpdate, patchedBytes);
-            } else {
-                sendRpcRequest(new TwinRequest(twinUpdate.getKey(), twinUpdate.getLocation()));
-            }
-        }
-    }
-
-    private byte[] applyPatch(TwinTracker twinTracker, TwinUpdate twinUpdate) {
-        if (!twinUpdate.isPatch()) {
-            return twinUpdate.getObject();
-        }
-        try {
-            JsonNode resultingDiff = objectMapper.readTree(twinUpdate.getObject());
-            JsonNode original = objectMapper.readTree(twinTracker.getObj());
-            JsonPatch patch = JsonPatch.fromJson(resultingDiff);
-            JsonNode resultNode = patch.apply(original);
-            return resultNode.toString().getBytes(StandardCharsets.UTF_8);
-        } catch (Exception e) {
-            LOG.error("Not able to apply patch for key {}", twinUpdate.getKey(), e);
-        }
-        return null;
-    }
-
-    private boolean isObjectUpdated(TwinTracker twinTracker, TwinUpdate twinUpdate) {
-        if (twinUpdate.isPatch()) {
-            return true;
-        }
-        return !Arrays.equals(twinTracker.getObj(), twinUpdate.getObject());
-    }
+        // Consume in thread instead of using broker's callback thread.
+        this.executorService.execute(() -> {
+            final var subscription = this.subscriptions.computeIfAbsent(twinUpdate.getKey(), Subscription::new);
 
-    private void updateSessions(TwinUpdate twinUpdate, byte[] newObjBytes) {
-        // Update twin object in local cache.
-        objMap.put(twinUpdate.getKey(), new TwinTracker(newObjBytes, twinUpdate.getVersion(), twinUpdate.getSessionId()));
-        // Send update to each session.
-        synchronized(sessionMap) {
-            final var sessions = sessionMap.get(twinUpdate.getKey());
-            if (sessions != null) {
-                sessions.forEach(session -> {
-                    try {
-                        session.accept(newObjBytes);
-                    } catch (Exception e) {
-                        LOG.error("Exception while sending update to Session {} for key {}", session, twinUpdate.getKey(), e);
-                    }
-                });
+            try {
+                subscription.update(twinUpdate);
+            } catch (final IOException e) {
+                LOG.error("Processing update failed: {}", twinUpdate.getKey(), e);
+                subscription.request();
             }
-        }
+        });
```

Comment:
The subscription could call `request` on itself instead of throwing an `IOException`.

Reply:
This would multiply the logging and the call to `request` to all places where an exception is thrown.
I foresee some more advanced error handling required in the future, which can happen in this exact spot, too.